src/deepsparse/transformers/pipelines/text_generation.py

```diff
@@ -0,0 +1,359 @@
+
+        :param engine_inputs: list of numpy inputs to
+            Pipeline engine forward pass
+        :return: A tuple of numpy array that contains the
+            sequence of generated tokens and a sequence
+            of logits for each generated token
+        """
+        # run the prompt through
+        tokens, logits = self.prompt_inference(engine_inputs)
+        num_prompt_tokens = len(tokens) - 1
+
+        # create the generated output
+        max_tokens = (
+            self.max_generated_tokens
+            if self.max_generated_tokens and self.max_generated_tokens > 0
+            else 100 * self.sequence_length
+        )  # set safety for absolute max generation
+
+        generated_tokens = [tokens[-1]]
+        generated_logits = [logits]
+
+        while len(generated_tokens) < max_tokens:
+            (
+                token,
+                logits,
+            ) = self.autoregressive_inference(tokens, num_prompt_tokens)
+            tokens.append(token)
+            generated_tokens.append(token)
+            generated_logits.append(logits)
+
+            if token == self.tokenizer.eos_token_id and not self.force_max_tokens:
+                break
+
+        return numpy.array([[generated_tokens]]), numpy.concatenate(
+            generated_logits, axis=1
+        )
+
+    def prompt_inference(
+        self, engine_inputs: List[numpy.ndarray]
+    ) -> Tuple[List[int], Dict[str, numpy.ndarray]]:
+        """
+        An inference run that processes the prompt through the
+        model to generate the new token and logits
+
+        :param engine_inputs: the prompt (context) represented by a
+            list of numpy inputs to the engine
+        :return: A tuple of:
+            - The list of prompt tokens plus the new, generated token
+            - The logits generated from the prompt (with dimensions
+            ['batch_size', 'num_tokens', 'vocab_size'])
+        """
+        # get tokens by attention mask
+        tokens = engine_inputs[0][engine_inputs[1].nonzero()].tolist()
+        new_token = None
+
+        if (
+            self.prompt_batch_threshold is None
+            or self.prompt_batch_threshold >= 1
+            or len(tokens) / float(self.sequence_length) < self.prompt_batch_threshold
+        ):
+            # prompt size is small, run autoregressive inference to populate kv cache
+            run_tokens = []
+            for token in tokens:
+                run_tokens.append(token)
+                new_token, new_logits = self.autoregressive_inference(
+                    run_tokens, num_prompt_tokens=0
+                )
+        else:
+            # larger prompt size, run through multi-token engine in single pass
+            new_token, new_logits = self.multitoken_engine(engine_inputs)
+
+        tokens.append(new_token)
+
+        return tokens, new_logits
+
+    def autoregressive_inference(
+        self,
+        tokens: List[int],
+        num_prompt_tokens: int,
```

Comment:
this argument can be removed, since now the decoder engine takes care of tracking that.